jupyter_client/manager.py

```diff
@@ -52,33 +52,38 @@ class _ShutdownStatus(Enum):
     SigkillRequest = "SigkillRequest"
 
 
-def in_pending_state(method):
+def in_pending_state(attr='_ready'):
     """Sets the kernel to a pending state by
     creating a fresh Future for the KernelManager's `ready`
     attribute. Once the method is finished, set the Future's results.
     """
 
-    @functools.wraps(method)
-    async def wrapper(self, *args, **kwargs):
-        # Create a future for the decorated method
-        try:
-            self._ready = Future()
-        except RuntimeError:
-            # No event loop running, use concurrent future
-            self._ready = CFuture()
-        try:
-            # call wrapped method, await, and set the result or exception.
-            out = await method(self, *args, **kwargs)
-            # Add a small sleep to ensure tests can capture the state before done
-            await asyncio.sleep(0.01)
-            self._ready.set_result(None)
-            return out
-        except Exception as e:
-            self._ready.set_exception(e)
-            self.log.exception(self._ready.exception())
-            raise e
+    def inner(method):
+        @functools.wraps(method)
+        async def wrapper(self, *args, **kwargs):
+            # Create a future for the decorated method
+            try:
+                asyncio.get_running_loop()
+                fut = Future()
+            except RuntimeError:
+                # No event loop running, use concurrent future
+                fut = CFuture()
```

Comment:
If this is inside an `async def`, can this except branch ever occur? There should always be a running loop from inside a coroutine.

The previous strategy of setting `_ready` _outside_ a coroutine might not have a running loop, but now that you've moved it into the coroutine (good, I think), there should always be a running loop.

Returning two types of Future depending on context would be tricky because they aren't interchangeable (CFuture is not awaitable without calling `asyncio.wrap_future(cfuture)`).

Reply:
An important difference of this strategy, though: the attribute will not be set until an arbitrarily-later point in time due to asyncness. The previous strategy ensured `._ready` was defined _immediately_ before any waiting, whereas putting it in the coroutine means the attribute will _not_ be defined immediately.

Two new things to consider (may well both be covered):

1. previous waits for `self._ready` that may now be called when `_ready` is not yet
2. can `_ready` be set and then re-set? If so, waits for `_ready` may get a 'stale' state before the new `_ready` future is attached. This can be avoided with a non-async `delattr` prior to the async bit.

Reply:
Originally `_ready` was only defined in the constructor, which is why we had the fallback option.  It seems like these futures are becoming problematic in general.  Maybe what we really want is a state and a signal when the state changes, like we have on the JavaScript side.

Reply:
Alternatively, if we want to define `ready` and `shutdown_ready` as "the most recent" starting and shutting down futures, we could also add an `is_ready` and `is_shutdown_ready` for when there are no pending of either.  We would use this trick that tornado used to work around the new deprecation warnings:

```python
loop = asyncio.get_event_loop_policy().get_event_loop()   # this always returns the same loop
future = aysncio.Future(loop=loop)
```

We could then make `.ready` and `.shutdown` ready promises only available when using the `AsyncKernelManager`.

Reply:
I think part of the problem is that `jupyter_client` is a library, not an application, in the same way that `tornado` is.   It looks like we also can't rely on `asyncio_run` in the future, since it relies on `get_event_loop` and `set_event_loop`.  I think here's a sketch of what we need to do:

- The base classes will be asynchronous and call the the private async versions of functions directly as needed.  We will only use futures and tasks from within async methods.
- The synchronous classes will work by wrapping the asynchronous public methods using a decorator that runs a new private asyncio loop as `return asyncio.new_event_loop().run_until_complete(async_method)`.
- We remove the `.ready` future and add new methods called `async def wait_for_pending_start()` and `async def wait_for_pending_shutdown()` to handle pending kernels.  
 
With this new structure, the synchronous managers could also support pending kernels, breaking it into two synchronous steps.

Reply:
We can handle the ready future in this PR and defer the other parts to a separate refactor PR.

Reply:
btw, concurrent.futures are generally more compatible and flexible than asyncio futures - they are threadsafe, reusable across loops, etc. If you always use a cf.Future, you can await it in a coroutine with `asyncio.wrap_future(cf_future)`. That may be the simplest solution here. It's the _inconsistency_ that I saw as a problem (`self._ready` _may_ be awaitable). Something like:

```python
# earlier, _not_ in a coroutine
self._ready = cf.Future()

async def wait_for_ready(self):
    if self._ready is None:
        raise RuntimeError("haven't started yet!")
    # wrap cf.Future self._ready to make it awaitable
    await asyncio.wrap_future(self._ready)
```

ought to be more reliable.


To always use new event loops (or `asyncio.run`) for each sync method call, one important consideration is that all your methods completely resolve by the end of the coroutine (i.e. not hold references to the current loop for later calls, which means requiring [pyzmq 22.2](https://pyzmq.readthedocs.io/en/v22.2.0/changelog.html#id1), among other things). A possibly simpler alternative is to maintain your _own_ event loop reference. `asyncio.get_event_loop()` is deprecated because it did too many things, but that doesn't mean _nobody_ should hold onto persistent loop references. It just means that they don't want to track a _global_ not-running loop. If each instance had a `self._loop = asyncio.new_event_loop()` and always ran the sync wrappers in that loop (`self._loop.run_until_complete(coro())`), it should work fine. That avoids the multiple-loop problem for sync uses because it's still a single persistent loop. Each instance may happen to use a different loop, but that should not be a problem in the sync case.

Reply:
Sweet, thanks for the insights Min.  That sounds like a good approach.

Reply:
More thoughts on ready future handling:

- Remove the public ready attribute(s).
- Internally, store current ready and pending flags for start and shutdown.
- In `start()`, check for `_start_pending` and error if true.  Set `_start_pending` and set a new `_start_ready` future and clear `_start_pending` when ready.
- In `shutdown()`, check for `_shutdown_pending` and bail if true.  Set `_shutdown_pending` and set a new `_shutdown_ready` future.  Clear `_shutdown_pending` when ready.
- In `wait_for_start_ready()` and `wait_for_shutdown_ready()` store current ready, and wait for it.  If new current ready is different from the stored one, recurse.